server/modules/monitor/mysql_mon.c

```diff
@@ -1127,27 +1127,28 @@ static void set_master_heartbeat(MYSQL_MONITOR *handle, MONITOR_SERVERS *databas
     }
 
     /* create the maxscale_schema database */
-    if (mysql_query(database->con, "CREATE DATABASE IF NOT EXISTS maxscale_schema"))
-    {
-        MXS_ERROR("[mysql_mon]: Error creating maxscale_schema database in Master server"
-                  ": %s", mysql_error(database->con));
-
+    if (mysql_query(database->con, "SELECT table_name FROM information_schema.tables WHERE table_schema = 'maxscale_schema' AND table_name = 'replication_heartbeat'")) {
+        MXS_ERROR( "[mysql_mon]: Error checking for replication_heartbeat in Master server"
+                ": %s", mysql_error(database->con));
         database->server->rlag = -1;
     }
-
-    /* create repl_heartbeat table in maxscale_schema database */
-    if (mysql_query(database->con, "CREATE TABLE IF NOT EXISTS "
-                    "maxscale_schema.replication_heartbeat "
-                    "(maxscale_id INT NOT NULL, "
-                    "master_server_id INT NOT NULL, "
-                    "master_timestamp INT UNSIGNED NOT NULL, "
-                    "PRIMARY KEY ( master_server_id, maxscale_id ) ) "
-                    "ENGINE=MYISAM DEFAULT CHARSET=latin1"))
+        
+    if (0 == mysql_num_rows(mysql_store_result(database->con)))
```

Comment:
This leaks memory since `mysql_store_result` allocates memory. It can also fail and the call to mysql_num_rows with a NULL value shouldn't be done.